python/pyarrow/parquet.py

```diff
@@ -925,7 +925,6 @@ def __init__(self, path_or_paths, filesystem=None, schema=None,
             self.paths = _parse_uri(path_or_paths)
 
         self.memory_map = memory_map
-        self._open_file_func = self._get_open_file_func()
```

Comment:
@pitrou I've reduced `self._open_file_func` to be an instancemethod instead of a closure, but still it's bound to self.

Reply:
I don't understand what you mean. What is the error you get on Python 2 ?

Reply:
https://travis-ci.org/apache/arrow/jobs/520755777#L3184

Reply:
I see, the following change should work:
```diff
diff --git a/python/pyarrow/parquet.py b/python/pyarrow/parquet.py
index ade0ddc79..966ae4a58 100644
--- a/python/pyarrow/parquet.py
+++ b/python/pyarrow/parquet.py
@@ -864,6 +864,13 @@ def _path_split(path, sep):
     return head, tail
 
 
+def _open_dataset_file(dataset, path, meta=None):
+    """
+    A Python 2-picklable function that just calls ParquetDataset._open_file.
+    """
+    return dataset._open_file(path, meta)
+
+
 EXCLUDED_PARQUET_PATHS = {'_SUCCESS'}
 
 
@@ -931,7 +938,7 @@ class ParquetDataset(object):
          self.common_metadata_path,
          self.metadata_path) = _make_manifest(
              path_or_paths, self.fs, metadata_nthreads=metadata_nthreads,
-             open_file_func=self._open_file)
+             open_file_func=partial(_open_dataset_file, self))
 
         if self.common_metadata_path is not None:
             with self.fs.open(self.common_metadata_path) as f:
```

Reply:
Ohh, didn't know that partial is supported. Trying it.